NodeJS EA QS (#8355)
auth0/docs: articles/quickstart/webapp/nodejs-ea/01-login.md

---
title: Login
description: This tutorial demonstrates how to add user login, logout, and profile to a Node.js Express application.
budicon: 448
topics:
  - quickstarts
  - webapp
  - login
  - user profile
  - logout
  - nodejs
contentType: tutorial
useCase: quickstart
github:
  path: 01-Login
---

<%= include('../../../_includes/_new_app', { hideDownloadSample: true, hideDashboardScreenshot: true }) %>

> <%= include('../../../_includes/_callback_url') %>

Reviewer: We should tell them what callback URL they're using (`/callback`) or they won't know.

> <%= include('../../../_includes/_logout_url' }) %>

Reviewer: We should tell them what logout URL they're using (`/` I believe) or they won't know.

## Integrate Auth0
### Install Dependencies
Install the following dependencies

- [`express-openid-connect`](https://github.com/auth0/express-openid-connect) - Auth0-maintained OIDC-compliant library for Express.

> - [`express-session`](https://github.com/auth0/express-session) - simple session middleware for Express (any session middleware for express will work).

Reviewer: ```suggestion
- [`express-session`](https://github.com/auth0/express-session) - session middleware for Express.
```

```sh

> npm install express-openid-connect express-session

Reviewer: ```suggestion
npm install express express-openid-connect express-session
```

Oddly, neither package installs `express` with this command 🤷‍♂

```

### Handling server responses
Your application will need to parse URL-encoded data sent back from the Auth0 server.  Express provides a middleware for this called `express.urlencoded`. If you are integrating an existing application that uses `urlencoded` from the `body-parser`module, that will work as well.

```js
const express = require('express');
const app = express();

app.use(express.urlencoded({
  extended: false
}));
```

### User sessions

> In order to keep the user logged in across requests you will need to use session middleware.  The session will allow you to recognize the user on concurrent requests. In this example, you will use `express-session` which stores the session in memory.

Reviewer: ```suggestion
To keep the user logged in across multiple requests, your application will use the `express-session` middleware, which stores the user's data in memory. The user's session will allow your application to recognize them from request to request.
```

```js
const session = require('express-session');

app.use(session({
  secret: 'this should be a secret',
  resave: true,
  saveUninitialized: false
}));
```

### Authentication

> The last part of integrating Auth0 into your application will be using the `auth` middleware provided by Auth0's `express-openid-connect` installed above.  View the official [API Documentation](https://github.com/auth0/express-openid-connect/blob/master/API.md) for configuration options and information.

Reviewer: ```suggestion
The last part of integrating Auth0 into your application will be using the `auth` middleware provided by the Express OpenID Connect library installed above. The [API documentation](https://github.com/auth0/express-openid-connect/blob/master/API.md) for this library outlines additional configuration options and methods.
```

```js
const { auth } = require('express-openid-connect');

app.use(auth({
  required: false,
  auth0Logout: true,
  baseURL: 'http://localhost:3000',
  issuerBaseURL: 'https://${account.namespace}',
  clientID: '${account.clientId}'
}));
```

> Now that your application has the middleware installed, you can use [various helpers and values](https://github.com/auth0/express-openid-connect/blob/master/API.md#session-and-context) provided by the `express-openid-connect` library.  Here is an example of using `req.openid.isAuthenticated`

Reviewer: ```suggestion
Now that your application has the middleware installed, you can use [the helper functions and contextual data](https://github.com/auth0/express-openid-connect/blob/master/API.md#session-and-context) provided by the Express OpenID Connect library.  Here is an example of how to use `req.openid.isAuthenticated()`:
```

```js
app.get('/', (req, res) => {

>  res.send('You are logged ' + req.openid.isAuthenticated() ? 'In' : 'Out')

Reviewer: ```suggestion
 res.send(req.isAuthenticated() ? 'Logged in' : 'Logged out'))
```

});
```

## Login

> To log a user in visit the `/login` route provided by the library, or if you are testing your application locally [`localhost:3000/login`](http://localhost:3000/)

Reviewer: ```suggestion
A user can log into your application by visiting the `/login` route provided by the library. If you are testing your application locally, that link is [`localhost:3000/login`](http://localhost:3000/login).
```

## Profile

> To retrieve the user profile you will first want to protect the route where you display that users information.

Reviewer: ```suggestion
To display the user's profile, your application should provide a protected route.
```

Add the `requiresAuth` middleware for routes that require authentication.  Any route using this middleware will check for a valid user session and, if one does not exist, it will redirect the user to log in.

```js
const { requiresAuth } = require('express-openid-connect');

app.get('/profile', requiresAuth(), (req, res) => {

>   // when a user is logged in you can access the user object via `req.openid.user`

Reviewer: ```suggestion
```

  res.send(JSON.stringify(req.openid.user));
});
```

## Logout

> To log a user out visit the `/logout` route provided by the library, or if you are testing your application locally [`localhost:3000/logout`](http://localhost:3000/)

Reviewer: ```suggestion
A user can log out of your application by visiting the `/logout` route provided by the library. If you are testing your application locally, that link is [`localhost:3000/logout`](http://localhost:3000/logout).
```

## What's next?

> The Auth0 team has put together a few different examples of using [`express-openid-connect`](https://github.com/auth0/express-openid-connect)

Reviewer: ```suggestion
We put together a few examples of how to use [Express OpenID Connect](https://github.com/auth0/express-openid-connect) in more advanced use cases:
```

---
title: Login
description: This tutorial demonstrates how to add user login, logout, and profile to a Node.js Express application.
budicon: 448
topics:
  - quickstarts
  - webapp
  - login
  - user profile
  - logout
  - nodejs
contentType: tutorial
useCase: quickstart
github:
  path: 01-Login
---

<%= include('../../../_includes/_new_app', { hideDownloadSample: true, hideDashboardScreenshot: true }) %>
<%= include('../../../_includes/_callback_url') %>
<%= include('../../../_includes/_logout_url' }) %>

## Integrate Auth0
### Install Dependencies
Install the following dependencies

> - [`express-session`](https://github.com/auth0/express-session) - simple session middleware for Express (any session middleware for express will work).

Reviewer: Let's switch this to `cookie-session`. Simpler to use and less "gotchas." Similar philosophy to why we're going to skip secret management. A few instances below as well.

Reviewer: Also ... it looks like the only two options are `cookie-session` and `express-session` so maybe just call out the ability to use `express-session` but may need additional configuration considerations.

> - [`express-openid-connect`](https://github.com/auth0/express-openid-connect) - an [OIDC](/protocols/oidc) compliant authentication library.

Reviewer: ```suggestion
- [`express-openid-connect`](https://github.com/auth0/express-openid-connect) - Auth0-maintained OIDC-compliant library for Express.
```

```sh
# Using NPM
npm install express-openid-connect express-session

# Using Yarn
yarn add express-openid-connect express-session
```

### Handling server responses

> Your application will need to parse url-encoded payloads sent back from the Auth0 server.  Express provides a middleware for this `express.urlencoded`. You might already be using `body-parser`, this will work too.

Reviewer: ```suggestion
Your application will need to parse URL-encoded data sent back from the Auth0 server.  Express provides a middleware for this called `express.urlencoded`. If you are integrating an existing application that uses `urlencoded ` from the `body-parser`module, that will work as well.
```

```js
const express = require('express');
const app = express();

app.use(express.urlencoded({
  extended: false
}));
```

### User sessions

> In order to keep the user logged in across requests you will need to use session middleware.  The session will allow you to recognize the user on concurrent requests. In this example, you will use `express-session` which stores the session in memory.

Reviewer: `cookie-session` update here as well.

```js
const session = require('express-session');

app.use(session({
  secret: 'this should be a secret',
  resave: true,
  saveUninitialized: false
}));
```

### Authentication

> The last part of integrating Auth0 into your application will be using the `auth` middleware supplied to you from the Auth0 library `express-opeinid-connect`.  View the official [Api Documentation](https://github.com/auth0/express-openid-connect/blob/4374c0502d4aedf8b3975e7c5f26929b305b32f6/API.md) for configuration options and information.

Reviewer: ```suggestion
The last part of integrating Auth0 into your application will be using the `auth` middleware provided by Auth0's `express-openid-connect` installed above.  View the official [API Documentation](https://github.com/auth0/express-openid-connect/blob/master/API.md) for configuration options and information.
```

```js
const { auth } = require('express-openid-connect');

app.use(auth({
  required: false,
  auth0Logout: true,
  baseURL: 'http://localhost:3000',
  issuerBaseURL: 'https://${account.namespace}',
  clientID: '${account.clientId}'
}));
```

## Protect a Route
Now that your application has the required middleware to integrate Auth0 in your application, you can start protecting routes.

Add the `requiresAuth` middleware supplied from `express-openid-connect` for routes that require authentication.  Any route using this middleware will check for a valid user session and, if one does not exist, it will redirect the user to log in.

```js
const { requiresAuth } = require('express-openid-connect');

app.get('/profile', requiresAuth(), (req, res) => {
  // when a user is logged in you can access the user object via `req.openid.user`
  res.send(JSON.stringify(req.openid.user));
});
```

> ## Login and Logout

Reviewer: Maybe instead of a main section header for both, Login and Logout get their own sections.

Author: pretty small sections, but we let's test it

### Login
To log a user in you have 3 options:
- Redirect to the `/login` route registered by the `express-oidc-connect.auth` middleware.
- Use the `express-oidc-connect.requiresAuth` middleware to protect a route, as described above.
- When initializing the `express-oidc-connect.auth` middleware, pass in `required: true` to force authentication on all routes.

::: note
If you are testing your application locally, you can login by visiting `http://localhost:3000/login`
:::

> 

Reviewer: I think we should add a **Profile** or **Get User Data** section here to call out `req.openid.user`. Maybe that's just a matter of renaming "Protect a Route" above and moving it here?

### Logout
To log a user out, redirect them to the `/logout` route registered by the `express-oidc-connect.auth` middleware.

::: note

> If you are testing your application locally, you can logout by visiting `http://localhost:3000/logout`

Reviewer: ```suggestion
If you are testing your application locally, you can logout by visiting [`localhost:3000/logout`](http://localhost:3000/logout)
```

I would combine this with the sentence above instead of calling it out (IMHO)

:::

> ## What's next?

Reviewer: It feels like we should call out specific developer tasks here, like what is in EXAMPLES:

https://github.com/auth0/express-openid-connect/blob/master/EXAMPLES.md

---
title: Login
description: This tutorial demonstrates how to add user login, logout, and profile to a Node.js Express application.
budicon: 448
topics:
  - quickstarts
  - webapp
  - login
  - user profile
  - logout
  - nodejs
contentType: tutorial
useCase: quickstart
github:
  path: 01-Login
---

<%= include('../../../_includes/_new_app', { hideDownloadSample: true, hideDashboardScreenshot: true }) %>

<%= include('../../../_includes/_callback_url') %>

:::note

> If you are running this project on your local machine, the callback URL you need to whitelist in the **Allowed Callback URLs** field is `http://localhost:3000/callback`.

Reviewer: ```suggestion
If you are running this project on your local machine, your application's callback URL is `http://localhost:3000/callback`
```

Reviewer: I don't think we need to partially duplicate the information above.

:::

<%= include('../../../_includes/_logout_url' }) %>

:::note

> If you are running this project on your local machine, the logout URL you need to whitelist in the **Allowed Logout URLs** field is `http://localhost:3000/logout`.

Reviewer: ```suggestion
If you are running this project on your local machine, your application's logout URL is `http://localhost:3000`
```

Reviewer: This is the URL that users come back to after logging out. Application setting field is a bit misleading ... same feedback as the above RE: duplicating steps

:::

## Integrate Auth0
### Install Dependencies

> Install the following dependencies

Reviewer: ```suggestion
Your application will need the following packages:
```

- [`express-openid-connect`](https://github.com/auth0/express-openid-connect) - Auth0-maintained OIDC-compliant library for Express.
- [`express-session`](https://github.com/auth0/express-session) - session middleware for Express.

```sh
npm install express express-openid-connect express-session
```

### Handling server responses
Your application will need to parse URL-encoded data sent back from the Auth0 server.  Express provides a middleware for this called `express.urlencoded`. If you are integrating an existing application that uses `urlencoded` from the `body-parser`module, that will work as well.

```js
const express = require('express');
const app = express();

app.use(express.urlencoded({
  extended: false
}));
```

### User sessions

> To keep the user logged in across multiple requests, your application will use the `express-session` middleware, which stores the user's data in memory. The user's session will allow your application to recognize them from request to request.

Reviewer: ```suggestion
To keep the user logged in across multiple requests, your application will use the `express-session` middleware to store the user's data in memory. The user's session will allow your application to recognize them from request to request.
```

```js
const session = require('express-session');

app.use(session({
  secret: 'this should be a secret',
  resave: true,
  saveUninitialized: false
}));
```

### Authentication

> The last part of integrating Auth0 into your application will be using the `auth` middleware provided by the Express OpenID Connect library installed above. The [API documentation](https://github.com/auth0/express-openid-connect/blob/master/API.md) for this library outlines additional configuration options and methods.

Reviewer: ```suggestion
The last part of integrating Auth0 into your application will be using the `auth` middleware provided by the Express OpenID Connect library installed above. The [API documentation](https://github.com/auth0/express-openid-connect/blob/master/API.md) for this library outlines additional configuration options and methods that can be used when initializing this middleware.
```

<%= include('../../../_includes/_logout_url' }) %>

## Integrate Auth0
### Express Middleware
There are three middlewares required to integrate Auth0 in your application.

- `express.urlencoded` - parses urlencoded payloads sent back from the Auth0 server.
- `express-session` - simple session middleware for Express (this can be swapped with any session middleware for express).
- `express-openid-connect.auth` - an [OIDC](/protocols/oidc) compliant authentication middleware.

::: note
To learn more about using middleware with Express view the [official documentation](https://expressjs.com/en/guide/using-middleware.html)
:::

```js
const express = require('express');
const session = require('express-session');
const { auth } = require('express-openid-connect');

// Example initialization of Express
const app = express();

// parses urlencoded payloads sent back from the Auth0 server
app.use(express.urlencoded({
  extended: false
}));

// session middleware required to handle user sessions
app.use(session({
  secret: 'this should be a secret',
  resave: true,
  saveUninitialized: false
}));

// express-openid-connect.auth
app.use(auth({

>   required: false,

Reviewer: I meant to ask earlier ... what do you think about this? The SDK is default "require auth for all routes." Do you think that's the most common case or do you think that applications would only need specific routes protected? If this default was switched, it feels like the DX is nicer as well:

```js
////
/// Auth required everywhere
//
const { auth } = require('express-openid-connect');

app.use(auth({ 
    required: true 
});

app.get('/profile', (req, res) => {
    // ...
});

////
/// Auth required on specific routes
//
const { auth, requiresAuth } = require('express-openid-connect');
app.use(auth());
app.get('/profile', requiresAuth(), (req, res) => {
    // ...
});
```

Reviewer: It's a fairly minor point but changing this is a pretty big breaking change and one I'd like to get in before EA if we decide to go that direction).

Reviewer: Going to punt this to beta. Have not had any complaints or comments about this, just something I was thinking about.

https://auth0team.atlassian.net/browse/SDK-1071

---
title: Login
description: This tutorial demonstrates how to add user login, logout, and profile to a Node.js Express application.
budicon: 448
topics:
  - quickstarts
  - webapp
  - login
  - user profile
  - logout
  - nodejs
contentType: tutorial
useCase: quickstart
github:
  path: 01-Login
---

<!-- markdownlint-disable MD002 -->
## Getting Started

> In this quickstart you will learn how to implement authentication in an Express application using the OIDC Compliant library [Express OIDC](https://github.com/auth0/express-openid-connect) authored and maintained by Auth0.

Reviewer: ```suggestion
In this quickstart you will learn how to implement authentication in an Express application using the [Express OpenID Connect](https://github.com/auth0/express-openid-connect) library, authored and maintained by Auth0.
```

::: warning

> Express OIDC is still in an [early access stage](/).

Reviewer: @albertoperdomo - Should we use generally the same messaging here as [on the repo](https://github.com/auth0/express-openid-connect#express-openid-connect)?

:::

### Install Dependencies

```sh
# Using NPM
npm install express-openid-connect express-session

# Using Yarn
yarn add express-openid-connect express-session
```

> ## Configure Auth0

Reviewer: I believe these should use includes like the [Passport one](https://raw.githubusercontent.com/auth0/docs/master/articles/quickstart/webapp/nodejs/01-login.md)

Author: I was trying to avoid all the boilerplate that comes loaded in those partials.  But addressing the length/brevity of that content could be addressed separately.

Reviewer: Fair enough. As long as it was intentional 👍

A callback URL is a URL in your application where Auth0 redirects the user after they have authenticated. 

The callback URL for your app must be whitelisted in the **Allowed Callback URLs** field in your [Application Settings](${manage_url}/#/applications/${account.clientId}/settings). If this field is not set, users will be unable to log in to the application and will get an error.

::: note
If you are testing your application locally, you should set the **Allowed Callback URL** to `http://localhost:3000/callback`
:::

### Configure Logout URLs

A logout URL is a URL in your application that Auth0 can return to after the user has been logged out of the authorization server. This is specified in the `returnTo` query parameter.

The logout URL for your app must be whitelisted in the **Allowed Logout URLs** field in your [Application Settings](${manage_url}/#/applications/${account.clientId}/settings). If this field is not set, users will be unable to log out from the application and will get an error.

::: note
If you are testing your application locally, the logout URL you need to whitelist in the **Allowed Logout URLs** field is `http://localhost:3000`.
:::

> ## Integrate Auth0

Reviewer: Same here, should probably use includes ... and should probably be above the last section

Author: In our user research we discovered people skipping this section in search for the first code block or how to install to get started.

Reviewer: Interesting! Well then, by golly, let's use the order that works!

### Auth0 Application Keys

You will need to retrieve some details for your application to communicate with Auth0. You can get these details from the [Application Settings](${manage_url}/#/applications/${account.clientId}/settings) section in the Auth0 dashboard.

You need the following information:

* **Domain**
* **Client ID**

> * **Client Secret**

Reviewer: We want to leave this out for now ... more explanation below.

### Express Middleware
There are three middlewares required to integrate Auth0 in your application.

> - `express.urlencoded` - parses urlencoded payloads sent back from the Auth0 server.

Reviewer: Might want to mention that `bodyParser.urlencoded` works the same and might already be in use in an application. It's also the way to get this working with a lower version of Express.

Author: Do we have more documentation on that?  What version of express does it stop being required?

Reviewer: The `urlencoded` always needs to be there but Express 4.16 added it to `express`. Before that version, you'd need to install `body-parser` and use that.

Author: The quickstart and the library call out that Express 4.16+ is required

Reviewer: True but both work and they might already be using body parser. I don't know if it hurts to have both in there or not but it might be worth it just to say that both are acceptable. Or maybe thats TMI?

> - `express-session` - simple session middleware for Express (other session middleware can be used as a replacement).

Reviewer: I don't think this is technically a middleware. Also, we probably just want to call out that there needs to be session handling of some kind. They can use `express-session` (stores in memory ... there was a conversation at some point internally about how this might not be ideal) or `cookie-session` or another session hander. We'll need to pick one for this QS (and be confident we're choosing the right one) but I think we should make it clear that either will work.

Author: Yes I noticed `cookie-session` was used in the internal doc, but used the more widely used library `express-session`.  We should pick one and stick with it.  Don't need to worry the user with choice here.

Reviewer: I agree ... but I want to make sure we're not *mandating* `express-session`. Cookie session works fine as well. Point being: there needs to be a session handler and the one we choose here is A or B. Just like env management in that sense.

Author: > but I think we should make it clear that either will work.

the wording `other session middleware can be used as a replacement` is possibly poor, but this was the intent

Author: @joshcanhelp lets revisit this after I chunk out this section

> - `express-openid-connect` - an [OIDC](/protocols/oidc) compliant authentication middleware.

Reviewer: Also not technically a middleware (it provides middlewares).

::: note
To learn more about using middleware with Express view the [official documentation](https://expressjs.com/en/guide/using-middleware.html)
:::

> ::: note

Reviewer: This is a good note to have but ... we've got "secrets" in the code below. In other QSes, we say something like this but integrate our recommended library into the QS.

Author: User research was showing that users wanted to handle secret management at a different stage and outside of the quickstart.  This is more of a PSA than anything.

Reviewer: Interesting ... so we're thinking that the environment stuff should be left out, besides a note like this? 

I think I'm biased because the whole "environment storage" was such a catch point for me early on in my dev experience. The big downside of including it in the QS is, of course, that this is duplicated across all QSes in this platform.

Author: The only reason for the note was because of sentiment shared internally that we need to signal to the user best practices in security, more specifically not to check secrets into the source code.  

Also discussed here https://github.com/auth0/docs/pull/8355/files/cd66335815abeb2f2bb399299c82f6408e5432cf#r333791381 possibly it is best to show the user how to manage secrets, but it does add an additional dependency and widens scope of what the user is doing in this tutorial

Secrets should never be checked into source control.  We recommend using environment variables in place of plain text.  You can use [Dotenv](https://github.com/motdotla/dotenv) to easily manage envrionment variables on your localhost.
:::

```js
const express = require('express');
const session = require('express-session');
const { auth } = require('express-openid-connect');

> 

Reviewer: I would actually skip the whole "middleware" summary above and parse this chunk of code out into smaller snippets with explanation in between.

Author: what is the benefit of having smaller code snippets?

Reviewer: It's more step-by-step, rather than a big chunk of explanation then a big chunk of code.

To be totally honest, though, maybe that's not what helps people the most.

Author: I say we work it into smaller chunks.  Just wanted some clarification on the motive

// parses urlencoded payloads sent back from the Auth0 server
app.use(express.urlencoded({
  extended: false
}));

> 

Reviewer: Split

// session middleware required to handle user sessions
app.use(session({
  secret: 'this should be a secret',
  resave: true,
  saveUninitialized: false
}));

> 

Reviewer: Split

// express-openid-connect.auth 
app.use(auth({

>   // required - forces authentication for all routes

Reviewer: The current setting of `false` does the opposite of forcing auth

Author: Passing `false` will make it not required, is that not clear?

Reviewer: As a setting, yes, but in this context, it sounds like you're saying that the value below forces auth, which it doesn't.

Author: As mentioned below, maybe comments in the code should be left for samples.

  required: false,

>   // baseURL - There is where your /callback, /login & /logout URLs will be registered

Reviewer: Confusing explanation here .... also, most of our quickstarts keep the code pretty clean and explain the values below. This makes it easy to copy/paste into an app without much modification.

Author: We could reword this, but inline comments don't change the difficulty of copy/pasting the code.

Reviewer: They do if you don't want the comments in your final product :)

  baseURL: 'http://localhost:3000',

>   // issuerBaseURL - This is your Auth0 Domain found in your Auth0 Application Settings

Reviewer: It's actually the URL version of that ... domain pre-pended with `https:`

>   issuerBaseURL: '${account.namespace}',

Reviewer: ```suggestion
  issuerBaseURL: 'https://${account.namespace}',
```

>   // clientID - This is your Auth0 Client ID found in your Auth0 Application Settings

Reviewer: ```suggestion
  // clientID - This is the Client ID found in your Auth0 Application settings
```

  clientID: '${account.clientId}',
  // clientSecret - This is your Auth0 Client Secret found in your Auth0 Application Settings

>   clientSecret: '${account.clientSecret}'

Reviewer: Leave this out as well

}));
```

> 

Reviewer: This is probably a good place to call out that these values can be stored in `process.env` ([ref](https://github.com/auth0/express-openid-connect#getting-started)) and picked up automatically in the library. It would cut down on the amount of code that's here.

Author: I was going back and forth on the amount of code versus too much magic.  The library offers these options,  but I was worrying that it introduces too much magic too soon with too little understanding.  For example, `issuerBaseURL` if this doesn't change for the user, what is the point of using `process.env` for it?  

I'm not against moving these to `process.env`, as it definitely cleans up the amount of code needed, but it introduces an additional dependency for this tutorial.

Reviewer: Apologies, I did not explain that very clearly. 

The `express-oidc-connect` library can pick up it's required config values from the environment automatically. [README has instructions on that](https://github.com/auth0/express-openid-connect#getting-started). So this QS would not get a new dependency, but it would be a bit of black magic and start to get into env vars. 

Since we're not using a client secret, configuring in the code is maybe not ideal but also not disastrous. All of these values are public (appear in the URL). So maybe leave as-is for simplicity and add a note like:

> For other configuration options and methods see the [docs](https://github.com/auth0/express-openid-connect/blob/master/API.md)

## Protect a Route
Now that your application has the required middleware to integrate Auth0 in your application, you can start protecting routes.

> Add the `requiresAuth` middleware supplied from `express-openid-connect` for routes that need protection.  Any time a route using this middleware is hit, it will check for a valid user session, if none exists it will ask the user to log in.

Reviewer: ```suggestion
Add the `requiresAuth` middleware supplied from `express-openid-connect` for routes that require authentication.  Any route using this middleware will check for a valid user session and, if one does not exist, it will redirect the user to log in.
```

```js
const { requiresAuth } = require('express-openid-connect');

> app.get('/logged_in_user', requiresAuth(), (req, res) => {

Reviewer: ```suggestion
app.get('/profile', requiresAuth(), (req, res) => {
```

  // when a user is logged in you can access the user object via `req.openid.user`

>   res.send(JSON.stringify(req.openid.user));

Reviewer: 😎

});
```

## Login and Logout
### Login
To log a user in you have 3 options:

> - Send a `GET` request to the `/login` route registered by the `express-oidc-connect.auth` middleware.

Reviewer: ```suggestion
- Redirect to the `/login` route registered by the `express-oidc-connect.auth` middleware.
```

> - Use the `express-oidc-connect.requiresAuth` middleware to protect a route.

Reviewer: Maybe move this to the top of the list also.

```suggestion
- Use the `express-oidc-connect.requiresAuth` middleware to protect a route, as described above.
```

> - When initializing the `express-oidc-connect.auth` middleware pass in `required: true` to force required authentication on all routes.

Reviewer: ```suggestion
- When initializing the `express-oidc-connect.auth` middleware, pass in `required: true` to force authentication on all routes.
```

::: note

> If you are testing your application locally, you can login by visiting `http://localhost:3000/login`

Reviewer: This is handy but I would combine it with the bullet point above.

Author: The purpose of putting it is in the note is to stay consistent with the Logged out user section, and we seem to use this format when informing the user of nice to knows or shortcuts.  If this note seems redundant, I think it makes more sense to remove it all together

Reviewer: I'm on the fence. The phrase:

> Redirect to the `/login` route registered by the `express-oidc-connect.auth` middleware.

... is accurate and clear to *me* but is that maybe too technical? I don't think it would hurt to add after that something like:

> You can login to this local application at [localhost:3000/login](http://localhost:3000/login).

:::

### Logged in Users

> Once a user is logged in it is attached to the `request` via the `req.openid` object. You can access the full user object via `req.openid.user`

Reviewer: 🤔 ... I'd reword this a bit. Using `request` here is confusing. Maybe something like:

> The user's profile is always stored in `req.openid.user`

This makes me wonder ... should we add some kind of helper like `req.openid.isAuthenticated()` to the SDK that returns boolean based on if `user` is empty or not?

Author: I wonder if this section was even needed, it is demonstrated how to access the user object in the protect a route section

Reviewer: You could be right. 

Also ... the method I mentioned already exists :)

https://github.com/auth0/express-openid-connect/blob/master/lib/context.js#L26-L28

### Logout

> To log a user out send a `GET` request to the `/logout` route registered by the `express-oidc-connect.auth` middleware.

Reviewer: ```suggestion
To log a user out, redirect them to the `/logout` route registered by the `express-oidc-connect.auth` middleware.
```
